wip: Workspace

gustavlrsn opened the pull request:
For:
- https://github.com/opencollective/opencollective/issues/6385

znarf commented on components/dashboard/Menu.js:
```diff
@@ -0,0 +1,373 @@
+          icon={<Cog size={16} />}
+          if={!isHost}
+          setExpanded={setExpanded}
+          section="COLL_SETTINGS"
```
strange name

znarf replied:
Could you look into this or explain the cryptic name?

gustavlrsn replied:
Pushed a fix to rename it to "SETTINGS".

It's a bit of a "pseudo" section in the sense that it only acts as a submenu with the section name only there to provide a way to keep track of what submenu/section is currently expanded, this could be refactored or simplified a bit.

Betree commented on components/dashboard/TopBar.js:
```diff
@@ -0,0 +1,198 @@
+/* eslint-disable styled-components-a11y/click-events-have-key-events */
```
Not a fan of hiding ESLint errors for an entire file; we tend to forget about them. I prefer:
- Hiding just for the concerned line(s)
- Disabling for an entire section of the code
- Or disabling the rule globally if it doesn't make sense for us

Not a blocker for me, but we should at least add a comment to explain why we're doing this so that the next developer working on this file we know why we did it.

gustavlrsn replied:
Agree. This file is actually just a copy of `/components/admin-panel/TopBar.js` (but importing other constants) - a bit of a temporary tech debt to support both the /dashboard and existing /admin-panel

Betree commented on components/dashboard/sections/Transactions.tsx:
```diff
@@ -0,0 +1,40 @@
+  const { transactions, account } = data || {};
+
+  return (
+    <React.Fragment>
+      <TransactionsPage
+        transactions={transactions}
+        account={account}
+        variables={variables}
+        LoggedInUser={LoggedInUser}
+        error={error}
+        loading={loading}
+        refetch={refetch}
+        router={router}
+        isDashboard
+      />
+    </React.Fragment>
```
:ice_cream: 

```suggestion
    <TransactionsPage
      transactions={transactions}
      account={account}
      variables={variables}
      LoggedInUser={LoggedInUser}
      error={error}
      loading={loading}
      refetch={refetch}
      router={router}
      isDashboard
    />
```

Betree commented on components/dashboard/Switcher.tsx:
```diff
@@ -0,0 +1,258 @@
+  activeSlug: PropTypes.string,
+};
+
+const Switcher = () => {
```
Not a really clear name, I would suggest. Also for the filename.

```suggestion
const AccountSwitcher = () => {
```

gustavlrsn replied:
Nice, fixed!

Betree commented on pages/signin.js:
```diff
@@ -90,10 +90,11 @@ class SigninPage extends React.Component {
 
   async initialize() {
-    if (this.props.token) {
+    const token = this.props.token || this.props.router.query.token;
```
I'm curious about why is this needed now?

gustavlrsn replied:
This makes the login links work on Vercel (to be able to test the workspace on preview deploys), but perhaps something to be removed and looked at in another PR. I'll remove it.

znarf reviewed (approved):
Looks good to me!

Betree reviewed (commented):
This is looking great! 

I wonder if we couldn't have taken a simpler approach for the engineering side: this new dashboard is basically the existing admin panel with an account switcher, a few added sections, and some UX/CSS on the menu. We could have tried and introduced the switcher, the sections then iterated on the design in separate PRs. We'll probably have the opportunity to discuss this during retrospectives. 

Anyway, it's there, it's working great, and the code looks clean so let's move forward! Some feedback below; feel free to iterate directly or document into an issue as you see fit. No blocker for me overall!

- "Manage Contributions" is missing a loading state (currently blank)
- Not sure if it's because I have a strange user profile, but I shouldn't see "Received" here
![image](https://user-images.githubusercontent.com/1556356/230294483-ae65da62-a9b2-4f23-a451-b84402751121.png)
- When opening the dashboard for a collective, the default view is "Contributions". This is not really useful as 99% of collectives don't have any recurring contributions set up. Would be good to keep the existing "Info" form as the default view. 
![image](https://user-images.githubusercontent.com/1556356/230301712-2c184c40-27e3-4fd5-a11e-b9eb7a8bbe3a.png)
- Only a few power-users will be affected, but when you have many entries in your menu this happens. Should have a max-height + scroll:
![Peek 2023-04-06 09-16](https://user-images.githubusercontent.com/1556356/230302793-ded9a660-1b7a-48df-96ff-bfff258e95c3.gif)
- This one should have a `pointer` cursor. It's also not focusable and can't be accessed with the keyboard. We'll need to fix accessibility before the public release.
![image](https://user-images.githubusercontent.com/1556356/230303155-863adce6-ffe3-4b48-a62b-6fe339cc403e.png)
- Can we sort the accounts by name? 
![image](https://user-images.githubusercontent.com/1556356/230303784-ef377901-804a-44a2-88df-33f50b80e220.png)
- I get duplicate entries in the switcher
![image](https://user-images.githubusercontent.com/1556356/230304408-ead3f6ea-24de-4e31-bc74-0d0dd2f59057.png)

gustavlrsn commented on components/dashboard/Menu.js:
```diff
@@ -0,0 +1,373 @@
+          icon={<Inbox size={16} />}
+          setExpanded={setExpanded}
+        >
+          Applications
```
Removed this, this was a leftover in testing simpler names in the menu "Applications" instead of "Pending Applications", but that can be something to consider in a future PR.

gustavlrsn commented on components/dashboard/Menu.js:
```diff
@@ -0,0 +1,373 @@
+            </React.Fragment>
+          }
+        >
+          Organization Settings
```
Fixed

kewitz commented on rewrites.js:
```diff
@@ -126,6 +126,14 @@ exports.REWRITES = [
+    destination: '/dashboard',
+  },
+  {
+    source: '/dashboard/:section?/:slug/:subpath*',
```
IMO, `/dashboard/:slug/:section?/:subpath*` is more intuitive.

gustavlrsn replied:
Yeah, I think I agree. I was a bit back and forth with this. 

The idea with putting the slug last was that the AccountSwitcher might eventually be optional and more of a filter, i.e. you could go to `/dashboard/expenses` and see all expenses that concern me regardless of which Collective they belong to.

But since that is just an idea at this stage, I'm inclined to change it to what you suggest.

kewitz commented on components/dashboard/Switcher.tsx:
```diff
@@ -0,0 +1,258 @@
+                    <Span truncateOverflow>{loggedInUserCollective?.name}</Span>
+                  </Flex>
+                </StyledMenuEntry>
+                {Object.entries(groupedMemberships).map(([collectiveType, memberships]) => {
```
We should filter out groups without memberships. This can result in an empty Archived group.

kewitz commented on components/host-dashboard/constants.js:
```diff
@@ -1,5 +1,5 @@
 export const HOST_SECTIONS = {
-  EXPENSES: 'expenses',
+  HOST_EXPENSES: 'host-expenses',
```
This breaks the existing admin panel. It displays the Expense page as "Conversations" for some reason:
![image](https://user-images.githubusercontent.com/2119706/230318689-247072e0-924a-4815-9d62-9ee23894710e.png)

gustavlrsn replied:
Ah, good catch! This was unintentional, did not consider that this component was outside of the forked /admin-panel folder

gustavlrsn replied:
Now fixed.

kewitz reviewed (changes requested):
Nice work, it looks awesome! Just make sure you fix the existing admin panel "Expense" section that is broken.

gustavlrsn commented:
> This is looking great!
> 
> I wonder if we couldn't have taken a simpler approach for the engineering side: this new dashboard is basically the existing admin panel with an account switcher, a few added sections, and some UX/CSS on the menu. We could have tried and introduced the switcher, the sections then iterated on the design in separate PRs. We'll probably have the opportunity to discuss this during retrospectives.
> 
> Anyway, it's there, it's working great, and the code looks clean so let's move forward! Some feedback below; feel free to iterate directly or document into an issue as you see fit. No blocker for me overall!

Thank you for the review!

Yes, it's definitely building on top of the existing admin panel, and I figured that iterating on that while also not messing with any existing functionality would be easier by simply forking the admin panel and later remove the legacy one when we are happy with the new version, to allow more rapid prototyping evolution and have it be a "beta"/early access feature. Would be very happy to discuss the pros/cons of this approach and how it can be done better.

I'll iterate a bit today on the UX issues before merging.

gustavlrsn commented:
> * "Manage Contributions" is missing a loading state (currently blank)

Fixed this.

> * Not sure if it's because I have a strange user profile, but I shouldn't see "Received" here
>   ![image](https://user-images.githubusercontent.com/1556356/230294483-ae65da62-a9b2-4f23-a451-b84402751121.png)
This is due to using the same expenses component for all account types to simplify things. 

Pushed a fix to disable the "Received" tab for individual accounts:
<img width="1300" alt="image" src="https://user-images.githubusercontent.com/1552194/230378424-cea72691-79e1-4802-9738-791cb59e713b.png">

Definitely not ideal UX either - we can iterate on this further, maybe removing the filter entirely and adjust the title of the page instead to indicate that it is Submitted expenses, similar to the existing "submitted-expenses" page, but I think it's a worthwhile effort to use the same component for all collective types.

> * When opening the dashboard for a collective, the default view is "Contributions". This is not really useful as 99% of collectives don't have any recurring contributions set up. Would be good to keep the existing "Info" form as the default view.

Definitely agree, but unsure of just pushing to Info instead, since it is in a submenu in the last item in the menu.

I think what we will pitch for next cycle and before a full release of the workspace is a "Home" page, and also a better "Contributions" page that makes sense for all collective types.

Maybe a quick fix right now is putting the Expenses page first and making that the default?

> * Only a few power-users will be affected, but when you have many entries in your menu this happens. Should have a max-height + scroll:

Fixed.

> * This one should have a `pointer` cursor. It's also not focusable and can't be accessed with the keyboard. We'll need to fix accessibility before the public release.

Fixed `pointer` and keyboard accessibility.

> * Can we sort the accounts by name?

Fixed.

> * I get duplicate entries in the switcher

Fixed.

gustavlrsn commented:
> Nice work, it looks awesome! Just make sure you fix the existing admin panel "Expense" section that is broken.

Thank you! Good catch, that is fixed.

Betree commented:
> Pushed a fix to disable the "Received" tab for individual accounts: <img alt="image" width="1300" src="https://user-images.githubusercontent.com/1552194/230378424-cea72691-79e1-4802-9738-791cb59e713b.png">
> Definitely not ideal UX either - we can iterate on this further, maybe removing the filter entirely and adjust the title of the page instead to indicate that it is Submitted expenses, similar to the existing "submitted-expenses" page, but I think it's a worthwhile effort to use the same component for all collective types.

I agree, disabling without explaining why it's disabled is not great. We should mimic `pages/submitted-expenses.tsx`. Not a blocker for me as long as we clearly document the problem somewhere.

> > * When opening the dashboard for a collective, the default view is "Contributions". This is not really useful as 99% of collectives don't have any recurring contributions set up. Would be good to keep the existing "Info" form as the default view.
> 
> Definitely agree, but unsure of just pushing to Info instead, since it is in a submenu in the last item in the menu.

Still a better default IMO.

> Maybe a quick fix right now is putting the Expenses page first and making that the default?

Legit alternative.

gustavlrsn commented:
> I agree, disabling without explaining why it's disabled is not great. We should mimic `pages/submitted-expenses.tsx`. Not a blocker for me as long as we clearly document the problem somewhere.

Good point regarding disabling filter options, reverted this and instead removed the filter and changed the page title for individual accounts.

> > > * When opening the dashboard for a collective, the default view is "Contributions". This is not really useful as 99% of collectives don't have any recurring contributions set up. Would be good to keep the existing "Info" form as the default view.
> > 
> > 
> > Definitely agree, but unsure of just pushing to Info instead, since it is in a submenu in the last item in the menu.
> 
> Still a better default IMO.
> 
> > Maybe a quick fix right now is putting the Expenses page first and making that the default?
> 
> Legit alternative.

Made `Expenses` the first/default page for now.
